src/scheduler/Executor.cpp

```diff
@@ -528,9 +529,29 @@ void Executor::threadPoolThread(int threadPoolIdx)
 
         // Execute the task
         int32_t returnValue;
+        bool migrated = false;
         try {
             returnValue =
               executeTask(threadPoolIdx, task.messageIndex, task.req);
+        } catch (const faabric::util::FunctionMigratedException& ex) {
+            SPDLOG_DEBUG(
+              "Task {} migrated, shutting down executor {}", msg.id(), id);
+
+            // Note that when a task has been migrated, we need to perform all
+            // the normal executor shutdown, but we must NOT set the result for
+            // the call.
+            migrated = true;
+            selfShutdown = true;
+            returnValue = -99;
+
+            // MPI migration
+            if (msg.ismpi()) {
+                // TODO - delete the pending migration
```

Comment:
At the moment, faabric does not know when the migration (app-wise) has finished. In fact, faabric does not even know when a migration is taking place. Faabric only is aware that some functions throw exceptions because they are are migrated.

The workaround I have is the following:
* When we call `MpiWorld::prepareMigration` the local leader sets a boolean flag in the MPI world (e.g. `hasBeenMigrated`).
* Everytime we hit a barrier, after _all_ ranks have hit the barrier (and before they have left it). The local leader checks for the `hasBeenMigrated` flag.
* If `true` remove the pending migration from the map.

This uses the fact that faasm _will_ call `prepareMigration` when migrating, but can't really be tested from within faabric (I have tested it using the distributed test in faasm).

I can't think of a not ad-hoc way of doing this due to the lack of information in faabric.